compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -420,132 +275,416 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
+     *   @return    optionally, if match can be reduced to a matching case: A pair of
+     *              bindings for all pattern-bound variables and the untyped RHS of the case.
+     */
+    def reduceTopLevelMatch(scrutinee: Tree, scrutType: Type, cases: List[untpd.CaseDef], typer: Typer)(implicit ctx: Context): MatchRedux = {
 
-      // The final expansion runs a typing pass over the inlined tree. See InlineTyper for details.
-      val expansion1 = InlineTyper.typed(expansion, pt)(inlineCtx)
+      val gadtSyms = typer.gadtSyms(scrutType)
 
-      /** All bindings in `bindingsBuf` except bindings of inlineable closures */
-      val bindings = bindingsBuf.toList.map(_.withPos(call.pos))
+      /** Try to match pattern `pat` against scrutinee reference `scrut`. If succesful add
+       *  bindings for variables bound in this pattern to `bindingsBuf`.
+       */
+      def reducePattern(bindingsBuf: mutable.ListBuffer[MemberDef], scrut: TermRef, pat: Tree): Boolean = {
+        val isImplicit = scrut.info == defn.ImplicitScrutineeTypeRef
+
+        def newBinding(name: TermName, flags: FlagSet, rhs: Tree): Symbol = {
+          val info = if (flags `is` Implicit) rhs.tpe.widen else rhs.tpe.widenTermRefExpr
+          val sym = newSym(name, flags, info).asTerm
+          bindingsBuf += ValDef(sym, constToLiteral(rhs))
+          sym
+        }
 
-      inlining.println(i"original bindings = $bindings%\n%")
-      inlining.println(i"original expansion = $expansion1")
+        def searchImplicit(name: TermName, tpt: Tree) = {
+          val evidence = typer.inferImplicitArg(tpt.tpe, tpt.pos)
+          evidence.tpe match {
+            case fail: Implicits.AmbiguousImplicits =>
+              ctx.error(typer.missingArgMsg(evidence, tpt.tpe, ""), tpt.pos)
+              true // hard error: return true to stop implicit search here
+            case fail: Implicits.SearchFailureType =>
+              false
+            case _ =>
+              if (name != nme.WILDCARD) newBinding(name, Implicit, evidence)
+              true
+          }
+        }
 
-      val (finalBindings, finalExpansion) = dropUnusedDefs(bindings, expansion1)
+        pat match {
+          case Typed(pat1, tpt) =>
+            if (isImplicit) searchImplicit(nme.WILDCARD, tpt)
+            else scrut <:< tpt.tpe && reducePattern(bindingsBuf, scrut, pat1)
+          case pat @ Bind(name: TermName, Typed(_, tpt)) if isImplicit =>
+            searchImplicit(name, tpt)
+          case pat @ Bind(name: TermName, body) =>
+            reducePattern(bindingsBuf, scrut, body) && {
+              if (name != nme.WILDCARD) newBinding(name, EmptyFlags, ref(scrut))
+              true
+            }
+          case Ident(nme.WILDCARD) =>
+            true
+          case pat: Literal =>
+            scrut.widenTermRefExpr =:= pat.tpe
+          case pat: RefTree =>
+            scrut =:= pat.tpe ||
+            scrut.widen =:= pat.tpe.widen && scrut.widen.classSymbol.is(Module) && {
```

Comment:
I would move the `.is(Module)` check before the `=:=` check since that's much more expensive